towhee/dataframe/dataframe.py

```diff
@@ -35,19 +37,171 @@ def __init__(self, name: str, data: List[Tuple[Variable]] = None):
                 `Dataframe`. These tuples can be interpreted as being direct outputs
                 into downstream operators.
         """
-        # self._iter = ScalarIterator(df)
         self._name = name
-        self._data = data
+        self._data = data if data is not None else []
+        self._registered_iter = []
+        self._start_index = 0
+        self._total = 0
+        self._end = False
+
+        self._lock = threading.Lock()
+        self._registered_lock = threading.Lock()
 
     @property
     def name(self):
         return self._name
 
-    def __getitem__(self, val):
-        return self.data[val]
+    def put(self, item: Tuple[Variable]) -> None:
+        assert not self._end, 'DataFrame %s already end, can not put data' % self._name
```

Comment:
Let's raise a `ValueError` or some other custom `DataframeError` (instead of `AssertionError`) here so that it is clearer to the user exactly what the issue is.

Reply:
We can define some custom errors after v0.1